daemon/main.go

```diff
@@ -718,9 +719,22 @@ func runDaemon() {
 
 	// Launch another cilium-health as an endpoint, managed by cilium.
 	log.Info("Launching Cilium health endpoint")
-	addressing := d.getNodeAddressing()
-	cancelHealth := health.LaunchAsEndpoint(d, addressing, option.Config.Opts)
-	defer cancelHealth()
+	controller.NewManager().UpdateController("cilium-health-ep",
+		controller.ControllerParams{
+			DoFunc: func() error {
+				addressing := d.getNodeAddressing()
+				if err = health.PingEndpoint(); err != nil {
+					return health.LaunchAsEndpoint(d, addressing)
+				}
```

Comment:
Could you add a comment explaining that we're using `PingEndpoint` as a proxy for whether the endpoint is running?
As alternatives, we can use a persistent boolean defined outside this function (it will be heap allocated automatically, and there is only one instance of this controller ever so no races occur). A more clear scheme would be to use a sync.Once. I'm not sure if you anticipate having to relaunch the healthendpoint more than once. If so, perhaps it's better yet to check that the pid is still running and is our child?

Reply:
Sure.

I figured that sync.Once could be used for this purpose, but it seemed more complicated than necessary here. It's most likely that any error that occurs while trying to contact the health ep is because of something that can be remedied by restarting the EP. Treating the first initialisation as just another such case means we don't have to single out the initialisation case and handle it differently. Also, even if this assumption is wrong, the controller model should isolate and represent the persistent error in the `cilium status` output nicely.

When Cilium is deployed as a systemd service, on upgrade we also want to kill the version of this endpoint that isn't the current child.